src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+             tzoffset = self.tzoffset, tzindex = 0 }
+end
+
+local parse_zone
+
+local function datetime_new_raw(epoch, nsec, tzoffset, tzindex)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset or 0
+    dt_obj.tzindex = tzindex or 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, fraction, offset)
+    local epochV = dt ~= nil and (builtin.tnt_dt_rdn(dt) - DT_EPOCH_1970_OFFSET) *
+                   SECS_PER_DAY or 0
+    local secsV = secs or 0
+    local fracV = fraction or 0
+    local ofsV = offset or 0
+    return datetime_new_raw(epochV + secsV - ofsV * 60, fracV, ofsV)
+end
+
+local function get_timezone(offset)
+    if type(offset) == 'number' then
+        return offset
+    elseif type(offset) == 'string' then
+        return parse_zone(offset)
+    end
+end
+
+-- create datetime given attribute values from obj
+local function datetime_new(obj)
+    if obj == nil or type(obj) ~= 'table' then
+        return datetime_new_raw(0, 0, 0)
+    end
+
+    local ymd = false
+    local hms = false
+    local dt = DT_EPOCH_1970_OFFSET -- default is 1970-01-01
+
+    local y = obj.year
+    if y ~= nil then
+        check_range(y, {1, 9999}, 'year')
+        ymd = true
+    end
+    local M = obj.month
+    if M ~= nil then
+        check_range(M, {1, 12}, 'month')
+        ymd = true
+    end
+    local d = obj.day
+    if d ~= nil then
+        check_range(d, {1, 31, -1}, 'day')
+        ymd = true
+    end
+    local h = obj.hour
+    if h ~= nil then
+        check_range(h, {0, 23}, 'hour')
+        hms = true
+    end
+    local m = obj.min
+    if m ~= nil then
+        check_range(m, {0, 59}, 'min')
+        hms = true
+    end
+    local s = obj.sec
+    if s ~= nil then
+        check_range(s, {0, 60}, 'sec')
+        hms = true
+    end
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    -- if there are separate nsec, usec, or msec provided then
+    -- timestamp should be integer
+    local int_ts = nsec ~= nil or usec ~= nil or msec ~= nil
+
+    local ts = obj.timestamp
+    local fraction
+    if ts ~= nil then
+        s, fraction = math_modf(ts)
+        if not int_ts then
+            nsec = fraction * 1e9
+        end
+        hms = true
+    end
+
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, {-720, 720}, offset)
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        y = y or 1970
+        M = M or 1
+        if d ~= nil and d < 0 then
+            d = builtin.tnt_dt_days_in_month(y, M)
+        end
+        dt = builtin.tnt_dt_from_ymd(y, M, d)
+    end
+
+    -- .hour, .minute, .second
+    local secs = 0
+    if hms then
+        secs = (h or 0) * 3600 + (m or 0) * 60 + (s or 0)
+    end
+
+    return datetime_new_dt(dt, secs, nsec, offset or 0)
+end
+
+local sz = 48
```

Comment:
Please, use normal names at least for global variables.

Reply:
Renamed to `TOSTRING_BUFSIZE`